regsync: runtime error: invalid memory address or nil pointer dereference

### Current Behavior

Given a certain config file, running `regsync  --config regsync.yaml check` yields following stack trace:

```text
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x7fb5f0]

goroutine 1 [running]:
github.com/regclient/regclient/internal/reghttp.(*clientResp).backoffSet(0xc0003465a0)
	github.com/regclient/regclient/internal/reghttp/http.go:667 +0xd0
github.com/regclient/regclient/internal/reghttp.(*clientResp).Next(0xc0003465a0)
	github.com/regclient/regclient/internal/reghttp/http.go:525 +0x4f3
github.com/regclient/regclient/internal/reghttp.(*Client).Do(0xc000304b80, {0x9e75d8?, 0xc000180500}, 0xc00031e860)
	github.com/regclient/regclient/internal/reghttp/http.go:239 +0x128
github.com/regclient/regclient/scheme/reg.(*Reg).RepoList(0xc000304b00, {0x9e75d8, 0xc000180500}, {0xc00032a2d0, 0x29}, {0xcfc440, 0x0, 0x8010102?})
	github.com/regclient/regclient/scheme/reg/repo.go:51 +0x450
github.com/regclient/regclient.(*RegClient).RepoList(0x9e16e8?, {0x9e75d8, 0xc000180500}, {0xc00032a2d0, 0x29}, {0xcfc440, 0x0, 0x0})
	github.com/regclient/regclient/repo.go:26 +0x9c
main.(*rootCmd).processRegistry(_, {_, _}, {{0xc00032a2d0, 0x29}, {0xc00032a330, 0x2c}, {0xc000318300, 0x8}, {{0xcfc440, ...}, ...}, ...}, ...)
	github.com/regclient/regclient/cmd/regsync/root.go:434 +0x1fb
main.(*rootCmd).process(_, {_, _}, {{0xc00032a2d0, 0x29}, {0xc00032a330, 0x2c}, {0xc000318300, 0x8}, {{0xcfc440, ...}, ...}, ...}, ...)
	github.com/regclient/regclient/cmd/regsync/root.go:405 +0x125
main.(*rootCmd).runCheck(0x0?, 0xc000308600, {0x0?, 0x0?, 0x0?})
	github.com/regclient/regclient/cmd/regsync/root.go:323 +0x145
github.com/spf13/cobra.(*Command).execute(0xc000308600, {0xc00031e1a0, 0x2, 0x2})
	github.com/spf13/cobra@v1.7.0/command.go:940 +0x87c
github.com/spf13/cobra.(*Command).ExecuteC(0xc000308000)
	github.com/spf13/cobra@v1.7.0/command.go:1068 +0x3a5
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v1.7.0/command.go:992
github.com/spf13/cobra.(*Command).ExecuteContext(...)
	github.com/spf13/cobra@v1.7.0/command.go:985
main.main()
	github.com/regclient/regclient/cmd/regsync/main.go:28 +0x147
```

### Expected Behavior

If a plain bug, then no stack trace, otherwise a proper error, explaining the possible mistake in the config file.

### Steps To Reproduce

Config file in use: 

```yaml
---
version: 1


creds:

  - registry: europe-west1-docker.pkg.dev
    pathPrefix: prj-XXX-XXX-XXX/mirror

defaults:

  backup: 'backup-{{.Ref.Tag}}-{{$t := time.Now }}{{printf "%d%d%d-%d%d" $t.Year $t.Month $t.Day $t.Hour $t.Minute}}'
  interval: 60m
  ratelimit:
    min: 100
    retry: 15m
  parallel: 2
  digestTags: true
  mediaTypes:
      ## default
      - "application/vnd.docker.distribution.manifest.v2+json"
      - "application/vnd.docker.distribution.manifest.list.v2+json"
      - "application/vnd.oci.image.manifest.v1+json"
      - "application/vnd.oci.image.index.v1+json"
      ## default
      #
      #
sync:

  - source: ghcr.io/external-secrets/external-secrets
    target: europe-west1-docker.pkg.dev/external-secrets
    type: registry
    repos:
      allow:
        - external-secrets
      deny: []
    tags:
      allow: []
      deny: []

  - source: docker-public.packages.atlassian.com/sox/atlassian
    target: europe-west1-docker.pkg.dev/atlassian/
    type: registry
    repos:
      allow: []
      deny: []
    tags:
      allow: []
      deny: []

  - source: quay.io/jetstack
    target: europe-west1-docker.pkg.dev/jetstack
    type: registry
    repos:
      allow:
        - cert-manager-cainjector
        - cert-manager-controller
        - cert-manager-ctl
        - cert-manager-webhook
      deny: []
    tags:
      allow:
        - v1.12.\\d+
      deny: []
```
### Version

```text
VCSTag:     v0.5.3
VCSRef:     63bcd1a551acbb357d6a6aea0c50fb193aa77da9
VCSCommit:  63bcd1a551acbb357d6a6aea0c50fb193aa77da9
VCSState:   clean
VCSDate:    2023-10-06T20:40:08Z
Platform:   linux/amd64
GoVer:      go1.21.1
GoCompiler: gc
```

### Environment

- using the downloaded binary from my $PATH
- Linux (Debian testing) AMD64
- multiple public registries as per config file

### Anything else

n/a

--- a/internal/reghttp/http.go
+++ b/internal/reghttp/http.go
@@ -707,6 +707,14 @@ func (c *Client) getHost(host string) *clientHost {
 		} else {
 			h.config = config.HostNewName(host)
 		}
+		// check for normalized hostname
+		if h.config.Name != host {
+			host = h.config.Name
+			hNormal, ok := c.host[host]
+			if ok && hNormal.initialized {
+				return hNormal
+			}
+		}
 	}
 	if h.auth == nil {
 		h.auth = map[string]auth.Auth{}

--- a/internal/reghttp/http_test.go
+++ b/internal/reghttp/http_test.go
@@ -1654,5 +1654,38 @@ func TestRegHttp(t *testing.T) {
 			t.Errorf("error closing request: %v", err)
 		}
 	})
+	t.Run("Host Normalized", func(t *testing.T) {
+		apiGet := map[string]ReqAPI{
+			"": {
+				Method:     "GET",
+				Repository: "project",
+				Path:       "manifests/tag-get",
+				Headers:    headers,
+				Digest:     getDigest,
+			},
+		}
+		getReq := &Req{
+			Host: tsHost + "/path",
+			APIs: apiGet,
+		}
+		resp, err := hc.Do(ctx, getReq)
+		if err != nil {
+			t.Errorf("failed to run get: %v", err)
+			return
+		}
+		if resp.HTTPResponse().StatusCode != 200 {
+			t.Errorf("invalid status code, expected 200, received %d", resp.HTTPResponse().StatusCode)
+		}
+		body, err := io.ReadAll(resp)
+		if err != nil {
+			t.Errorf("body read failure: %v", err)
+		} else if !bytes.Equal(body, getBody) {
+			t.Errorf("body read mismatch, expected %s, received %s", getBody, body)
+		}
+		err = resp.Close()
+		if err != nil {
+			t.Errorf("error closing request: %v", err)
+		}
+	})
 	// TODO: test various TLS configs (custom root for all hosts, custom root for one host, insecure)
 }

--- a/repo.go
+++ b/repo.go
@@ -2,6 +2,8 @@ package regclient
 
 import (
 	"context"
+	"fmt"
+	"strings"
 
 	"github.com/regclient/regclient/scheme"
 	"github.com/regclient/regclient/types"
@@ -15,6 +17,10 @@ type repoLister interface {
 // RepoList returns a list of repositories on a registry.
 // Note the underlying "_catalog" API is not supported on many cloud registries.
 func (rc *RegClient) RepoList(ctx context.Context, hostname string, opts ...scheme.RepoOpts) (*repo.RepoList, error) {
+	i := strings.Index(hostname, "/")
+	if i > 0 {
+		return nil, fmt.Errorf("invalid hostname: %s%.0w", hostname, types.ErrParsingFailed)
+	}
 	schemeAPI, err := rc.schemeGet("reg")
 	if err != nil {
 		return nil, err
@@ -24,5 +30,4 @@ func (rc *RegClient) RepoList(ctx context.Context, hostname string, opts ...sche
 		return nil, types.ErrNotImplemented
 	}
 	return rl.RepoList(ctx, hostname, opts...)
-
 }

--- a/repo_test.go
+++ b/repo_test.go
@@ -1,0 +1,33 @@
+package regclient
+
+import (
+	"context"
+	"errors"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/regclient/regclient/types"
+)
+
+func TestRepoList(t *testing.T) {
+	ctx := context.Background()
+	log := &logrus.Logger{
+		Out:       os.Stderr,
+		Formatter: new(logrus.TextFormatter),
+		Hooks:     make(logrus.LevelHooks),
+		Level:     logrus.WarnLevel,
+	}
+	delayInit, _ := time.ParseDuration("0.05s")
+	delayMax, _ := time.ParseDuration("0.10s")
+	rc := New(
+		WithLog(log),
+		WithRetryDelay(delayInit, delayMax),
+	)
+	_, err := rc.RepoList(ctx, "registry.example.com/path")
+	if !errors.Is(err, types.ErrParsingFailed) {
+		t.Errorf("RepoList unexpected error on hostname with a path: expected %v, received %v", types.ErrParsingFailed, err)
+	}
+}

--- a/scheme/reg/reg.go
+++ b/scheme/reg/reg.go
@@ -108,7 +108,15 @@ func (reg *Reg) hostGet(hostname string) *config.Host {
 	reg.muHost.Lock()
 	defer reg.muHost.Unlock()
 	if _, ok := reg.hosts[hostname]; !ok {
-		reg.hosts[hostname] = config.HostNewName(hostname)
+		newHost := config.HostNewName(hostname)
+		// check for normalized hostname
+		if newHost.Name != hostname {
+			hostname = newHost.Name
+			if h, ok := reg.hosts[hostname]; ok {
+				return h
+			}
+		}
+		reg.hosts[hostname] = newHost
 	}
 	return reg.hosts[hostname]
 }

--- a/scheme/reg/repo_test.go
+++ b/scheme/reg/repo_test.go
@@ -256,4 +256,20 @@ func TestRepo(t *testing.T) {
 		}
 		// error is a json error, no custom error type was made for this yet
 	})
+	t.Run("Normalize host", func(t *testing.T) {
+		u, _ := url.Parse(tss["registry"].URL)
+		host := u.Host
+		rl, err := reg.RepoList(ctx, host+"/path")
+		if err != nil {
+			t.Errorf("error listing repos: %v", err)
+			return
+		}
+		rlRepos, err := rl.GetRepos()
+		if err != nil {
+			t.Errorf("error retrieving repos: %v", err)
+		} else if stringSliceCmp(listRegistry, rlRepos) == false {
+			t.Errorf("repositories do not match: expected %v, received %v", listRegistry, rlRepos)
+		}
+	})
+
 }

